lib/util.js

```diff
@@ -282,612 +365,515 @@ inspect.styles = Object.assign(Object.create(null), {
-      if (i > 0 === false)
-        continue;
-      const emptyItems = i - index;
-      const ending = emptyItems > 1 ? 's' : '';
-      const message = `<${emptyItems} empty item${ending}>`;
-      output.push(ctx.stylize(message, 'undefined'));
-      index = i;
-      if (++visibleLength === maxLength)
+// The array is sparse and or has extra keys
+function formatSpecialArray(ctx, value, recurseTimes, keys, maxLength, valLen) {
+  const output = [];
+  const keyLen = keys.length;
+  var visibleLength = 0;
+  var i = 0;
+  if (keyLen !== 0 && numberRegExp.test(keys[0])) {
+    for (const key of keys) {
+      if (visibleLength === maxLength)
+        break;
+      const index = +key;
+      // Arrays can only have up to 2^32 - 1 entries
+      if (index > 2 ** 32 - 2)
         break;
+      if (i !== index) {
+        if (!numberRegExp.test(key))
+          break;
+        const emptyItems = index - i;
+        const ending = emptyItems > 1 ? 's' : '';
+        const message = `<${emptyItems} empty item${ending}>`;
+        output.push(ctx.stylize(message, 'undefined'));
+        i = index;
+        if (++visibleLength === maxLength)
+          break;
+      }
+      output.push(formatProperty(ctx, value, recurseTimes, key, 1));
+      visibleLength++;
+      i++;
     }
-    output.push(formatProperty(ctx, value, recurseTimes, visibleKeys,
-                               elem, true));
-    visibleLength++;
-    index++;
   }
-  if (index < value.length && visibleLength !== maxLength) {
-    const len = value.length - index;
+  if (i < valLen && visibleLength !== maxLength) {
+    const len = valLen - i;
     const ending = len > 1 ? 's' : '';
     const message = `<${len} empty item${ending}>`;
     output.push(ctx.stylize(message, 'undefined'));
-    index = value.length;
+    i = valLen;
+    if (keyLen === 0)
+      return output;
   }
-  var remaining = value.length - index;
+  const remaining = valLen - i;
   if (remaining > 0) {
     output.push(`... ${remaining} more item${remaining > 1 ? 's' : ''}`);
   }
-  for (var n = 0; n < keys.length; n++) {
-    var key = keys[n];
-    if (typeof key === 'symbol' || !numbersOnlyRE.test(key)) {
-      output.push(formatProperty(ctx, value, recurseTimes, visibleKeys,
-                                 key, true));
+  if (ctx.showHidden && keys[keyLen - 1] === 'length') {
+    // No extra keys
+    output.push(formatProperty(ctx, value, recurseTimes, 'length', 2));
+  } else if (valLen === 0 ||
+    keyLen > valLen && keys[valLen - 1] === `${valLen - 1}`) {
+    // The array is not sparse
+    for (i = valLen; i < keyLen; i++)
+      output.push(formatProperty(ctx, value, recurseTimes, keys[i], 2));
+  } else if (keys[keyLen - 1] !== `${valLen - 1}`) {
+    const extra = [];
+    // Only handle special keys
+    var key;
+    for (i = keys.length - 1; i >= 0; i--) {
+      key = keys[i];
+      if (numberRegExp.test(key) && +key < 2 ** 32 - 1)
```

Comment:
Ditto about removing use of regexp here and simply checking for `NaN`.

Reply:
Using NaN is not sufficient here, we have to use the RegExp as e.g. `+004 === 4` and that is wrong.

Reply:
Oh, and if I remember correct it was even a tad faster to use the RegExp.